### DISCUSSION TITLE
exporting colormap indexes in the C API

### DISCUSSION THREAD

**Discussion by dmhursh** [ImageMagick/ImageMagick#7550] [Help]
### ImageMagick version

7.1.1-34

### Operating system, version and so on

windows 10 (and hopefully others)

### Description

I can easily export RGB & grayscale values from images using `MagickExportImagePixels()` with a wand.

Going to the `Image` I can determine if an image has  a colormap with `Image->storage_class`.

To read the colormap I have to use `(unsigned char) Image->colormap[index].red / (MaxColormapSize / 256UL)` with `.red`, `.green` & `.blue`

Is there a way like `MagickExportImagePixels()` to export the index values? 

Under the covers `MagickExportImagePixels()` uses `ExportImagePixels()` which calls functions like `ExportShortPixels()` which ultimately calls *unexported* functions like `GetPixelRed()`.  There is a function `GetPixelIndex()` that is unexported like `GetPixelRed()`. 

Could functions like `ExportShortPixels()` be patched to support a `QuantumType` like `IndexQuantum` wwhen you pass a character in the `map` argument for `ExportImagePixels()`?

Actually, while writing this, it appears the `I`/`i` is getting confused for both intensity and index.  `ExportImagePixels()` treats `I` as `IndexQuantum`.  `ExportShortPixels()` catches the `I` and overrides with `GetPixelIntensity()`.

Can a case be added to `ExportImagePixels()` to treat `X` as `IndexQuantum`?

**Comment by snibgo** [comment]
`magick in.png -verbose info:` gives the colormap as text. You can extract that in a script. To do it in a program, you can see how that is done by examining identify.c and searching for "Colormap entries".

**Comment by dmhursh** [comment]
That's how I figured out how to get the color map.  I'll see if I can tell how it reads the color indexes from the image and report back.

**Comment by dmhursh** [comment]
As I look into this more, I'm curious what the difference is between `GrayQuantum` and `IndexQuantum`.

**Comment by dmhursh** [comment]
The code for "Colormap entries" dumps the map of indexes to RGBs.  I'm able to do that.  What I am tried to do is read the per-pixel index values into that map.

**Comment by snibgo** [comment]
>  What I am tried to do is read the per-pixel index values into that map.

I don't understand.

`image->colormap` is an array of structures. You can read or write its values.

**Comment by dmhursh** [comment]
I can and do read RGBs from  `image->colormap`.

I want to read a pixel in row X & column Y *in the image* with the colormap.  Instead of reading the pixel's RGB, I want to read it's index into the colormap.

In other words, I have an image with the color map
```
0 = 0xFFFFFF
1 = 0xFFFFFF
2 = 0x000000
3 = 0x080808
```
I can read RBGs or Intensities for a pixel X,Y with the color 3 as 0x08.  I want to get the index of 3.  Of note, I want to tell pixels with color 0 from pixels with color 1 which both have the same RGB.

Sorry if I'm not being clear.  I'm mixing terms from a few different code bases.  

Looking at the code, imagemagick will get the green channel for a pixel with `GetPixelGreen()` in `pixel-accessor.h`.  I think what I'm asking for is from `GetPixelIndex()`, but I don't think that's exported.

**Comment by snibgo** [comment]
`GetPixelIndex()`, `GetPixelGreen()` etc are defined in `pixel-accessor.h`. If your program has `#include <MagickCore/MagickCore.h>`, it can use those functions.

**Comment by dmhursh** [comment]
Ah.  I haven't done much with static inlines in C.  I'll try that out later today.  Do you know if there is a reason to not make the index channel available via `MagickExportImagePixels()`.  Think a patch to add it would be accepted?  I looks like it was intended, but it returns intensity instead.

**Comment by snibgo** [comment]
> Think a patch to add it would be accepted?

It sounds like a good idea to me, provided it doesn't break the current behaviour.

**Comment by dmhursh** [comment]
Would you be someone  to coordinate with?  I have a couple questions.

- What character would be good to use for the map argument?  "I" is used for intensity.  I noticed "X" is free which I thought of for a "cross reference".  Not really mnemonic.   In anotther code based we've called the color map a "lookup table" or "lut".  I could maybe use "L"?

- In the plumbing, `IndexQuantum` has been used to mean "grayscale" or "intensity".  I think that's wrong, but it can't be changed.  Should I create a new `QuantumType` for it  or should just peek in the map string down in the ` ExportShortPixels()` type functions?

Thanks for your patience.

**Comment by snibgo** [comment]
I have no authority over the code. (And I don't want that authority.)

I should have said: It sounds like a good idea to me, provided it doesn't break current *correct* behaviour. If the current behaviour is buggy, then that can and should change.

I don't have time to look into this, but you seem to have identified (a) one or more bugs, and (b) an enhancement. You might show a simple example that demonstrates the bug(s), and code that corrects the bug(s) and supplies the enhancement.
